src/hidra/receiver/asapo_transfer.py

```diff
@@ -0,0 +1,77 @@
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import argparse
+import logging
+
+from hidra import Transfer
+from plugins.asapo_producer import AsapoWorker
+
+logger = logging.getLogger(__name__)
+
+
+class AsapoTransfer:
+    def __init__(self, endpoint, beamtime, token, n_threads, file_regex,
+                 default_data_source=None, timeout=5, beamline='auto',
+                 start_file_idx=1):
+
+        self.signal_host = "localhost"
+        self.targets = [["localhost", "50101", 1]]
+        self.asapo_worker = AsapoWorker(endpoint, beamtime, token, n_threads, file_regex,
+                                        default_data_source, timeout, beamline,
+                                        start_file_idx)
+
+        self.start_transfer()
+
+    def start_transfer(self):
+
+        query = Transfer("STREAM_METADATA", self.signal_host)
+        query.initiate(self.targets)
+        query.start()
+        self.run(query)
+        query.stop()
+
+    def run(self, query):
+        while True:
+            try:
+                [metadata, data] = query.get()
+            except:
+                break
```

Comment:
Don't break the loop but try to recover (this might be quite hard, though).

Reply:
I put log message instead of breaking. Not sure, that it what you mean.

Reply:
Hidra errors are now caught in the outermost (main) loop, resulting in a complete re-initialization of the connection.  Other errors are now handled directly and cause messages to be skipped.